src/pymordemos/fin_optimization.py

```diff
@@ -0,0 +1,247 @@
+    RB_greedy_data = rb_greedy(fom, RB_reductor, training_set, atol=1e-2)
+    rom = RB_greedy_data['rom']
+
+    def rom_objective_functional(mu):
+        return rom.output(mu)[0, 0]
+
+    def rom_gradient_of_functional(mu):
+        return rom.output_d_mu(fom.parameters.parse(mu), return_array=True, use_adjoint=True)
+
+    opt_rom_minimization_data = {'num_evals': 0,
+                                 'evaluations': [],
+                                 'evaluation_points': [],
+                                 'time': np.inf,
+                                 'offline_time': RB_greedy_data['time']}
+
+    tic = perf_counter()
+    opt_rom_result = minimize(partial(record_results, rom_objective_functional,
+                                      opt_rom_minimization_data),
+                              initial_guess,
+                              method='L-BFGS-B',
+                              jac=rom_gradient_of_functional,
+                              bounds=(range_1, range_2, range_2, range_2, range_2, range_2),
+                              options={'ftol': 1e-15})
+    opt_rom_minimization_data['time'] = perf_counter()-tic
+
+    print("\nResult of optimization with FOM model and adjoint gradient")
+    report(opt_fom_result, fom.parameters.parse, opt_fom_minimization_data, reference_mu)
+    print("Result of optimization with ROM model and adjoint gradient")
+    report(opt_rom_result, fom.parameters.parse, opt_rom_minimization_data, reference_mu)
+
+
+def create_fom():
+    rng = new_rng(222)
+    functions = [
+        ExpressionFunction('(2.5 <= x[0]) * (x[0] <= 3.5) * (0 <= x[1]) * (x[1] <= 4) * 1.', dim_domain=2),
+        ExpressionFunction('(0 <= x[0]) * (x[0] < 2.5) * (0.75 <= x[1]) * (x[1] <= 1) * \
+            1. + (3.5 < x[0]) * (x[0] <= 6) * (0.75 <= x[1]) * (x[1] <=1) * 1.', dim_domain=2),
+        ExpressionFunction('(0 <= x[0]) * (x[0] < 2.5) * (1.75 <= x[1]) * (x[1] <= 2) * 1. \
+            + (3.5 < x[0]) * (x[0] <= 6) * (1.75 <= x[1]) * (x[1] <= 2) * 1.', dim_domain=2),
+        ExpressionFunction('(0 <= x[0]) * (x[0] < 2.5) * (2.75 <= x[1]) * (x[1] <= 3) *1. \
+            + (3.5 < x[0]) * (x[0] <= 6) * (2.75 <= x[1]) * (x[1] <= 3) * 1.', dim_domain=2),
+        ExpressionFunction('(0 <= x[0]) * (x[0] < 2.5) * (3.75 <= x[1]) * (x[1] <= 4) *1. \
+            + (3.5 < x[0]) * (x[0] <= 6) * (3.75 <= x[1]) * (x[1] <= 4) * 1.', dim_domain=2)
+    ]
+    coefficients = [
+        ProjectionParameterFunctional('k0', 1),
+        ProjectionParameterFunctional('k1', 1),
+        ProjectionParameterFunctional('k2', 1),
+        ProjectionParameterFunctional('k3', 1),
+        ProjectionParameterFunctional('k4', 1)
+    ]
+    diffusion = LincombFunction(functions, coefficients)
+    parameter_ranges = {
+        'biot': (.01, 1),
+        'k0': (.1, 10),
+        'k1': (.1, 10),
+        'k2': (.1, 10),
+        'k3': (.1, 10),
+        'k4': (.1, 10)
+    }
```

Comment:
for the 6d problem, this demo takes quite long (as expected) especially because you are constructing a global rom for it. Maybe go to the 2d-case from the Qian et al paper to be faster. In general, we want this demo to be as fast as possible while showing us all features.